[13.0][MIG][WIP] product_cost_price_avco_sync: Migration to 13.0

carlosdauden opened the pull request:
This module is used to maintain the stock valuation with correct data retroactively.
For example:
On 2021-06-28 merchandise is received, but someone wrongly entered the price or quantity in the purchase order, putting 1000 instead of 10.00.
On 2021-07-01 the error is detected and solved.
When taking a valuation report on 2021-06-30 the information will appear incorrect (although it will be correct on 2021-07-02)

If, in addition, we have deliveries made on 2021-06-29, the svl related with this delivery will be wrong forever.

What this module does is synchronize the corrections so that all the records that are affected by that modification are modified sequentially.

In combination with other modules, sales order margins can be correctly calculated.

The only "drawback" of using this module is that if a report is obtained by selecting the same date at two different times, the information obtained may be different.

@Tecnativa

sergio-teruel committed:
[ADD] product_cost_price_avco_sync: New module to sync product cost with stock move price unit

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Cover more use cases

carlosdauden committed:
[MIG] product_cost_price_avco_sync: Migration to 12.0

carlosdauden committed:
[FIX] product_cost_price_avco_sync: Aleatory error because compare value before BD write.

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Fix test. Aleatory error

marcelsavegnago committed:
Added translation using Weblate (Portuguese (Brazil))

fcolus committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (2 of 2 strings)

Translation: stock-logistics-workflow-12.0/stock-logistics-workflow-12.0-product_cost_price_avco_sync
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-12-0/stock-logistics-workflow-12-0-product_cost_price_avco_sync/pt_BR/

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Cover case if multiple move lines with same product and distinct price are done at same second.

sergio-teruel committed:
[MIG] product_cost_price_avco_sync: Migration to v13.0

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: black, isort

carlosdauden committed:
[MIG] product_cost_price_avco_sync: SVL compatibility

carlosdauden committed:
[MIG] product_cost_price_avco_sync: Simplify code (or not)

sergio-teruel committed:
[MIG] product_cost_price_avco_sync: Compute returned moves from outgoing svl

sergio-teruel committed:
[MIG] product_cost_price_avco_sync: Remove old price

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Update inventory adjust moves value

pedrobaeza commented:
Please check Travis

chienandalu commented:
> Please check Travis

I'd say it just fails in pre-commit as tests are commented :sweat_smile:

pedrobaeza commented:
@carlosdauden can you please finish this and re-activate tests?

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Better compute layers with zero updates quantities and amounts
TT29454

sergio-teruel committed:
[IMP] Improvements

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Flake. Improve code and remove unnecessary with_context

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Flake. Vaccum manual adjustement

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Inventory adjustments

sergio-teruel committed:
[FIX] product_cost_price_avco_sync: Division by zero

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Round decimals

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Test and cover use cases

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Cover case elaboration with inventory IN OUT moves

sergio-teruel committed:
[FIX] product_cost_price_avco_sync: Singleton on _create_correction_svl

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Keep inventory unit_cost if there is not previous in moves or manual adjustment

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Discard scrap and returned location as inventory moves

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Avoid write stock move quantity done if quantity is the same or the move has more than one stock move lines

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Discard inventory moves if have a picking linked

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Discard scrap stock moves

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Add context to keep quantity after inventory adjustment

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Sync after write

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Always compute remaining qty

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Use dict to avoid modify many records

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Remove unnecessary rounds.

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Remove update_enabled.

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Simplify code and black style

carlosdauden committed:
[FIX] product_cost_price_avco_sync: Value field are not in dict

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Compare unit_cost field with currency precision

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Remaining quantities

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Simplify code

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: Remove unneeded quantity in vals

sergio-teruel committed:
[IMP] product_cost_price_avco_sync: variable unit cost processed after first adjustment input

AaronHForgeFlow commented:
Hi @carlosdauden @sergio-teruel  Can you please let me know why is this module needed. Is it to cover the case where the product changes its cost method? or is there an inconsistency in standard Odoo in the process of computing the price? Thanks in advance!

carlosdauden commented:
Hi @AaronHForgeFlow 

This module is used to maintain the stock valuation with correct data retroactively.
For example:
  On 2021-06-28 merchandise is received, but someone wrongly entered the price or quantity in the purchase order, putting 1000 instead of 10.00.
  On 2021-07-01 the error is detected and solved.
  When taking a valuation report on 2021-06-30 the information will appear incorrect (although it will be correct on 2021-07-02)

If, in addition, we have deliveries made on 2021-06-29, the svl related with this delivery will be wrong forever.

What this module does is synchronize the corrections so that all the records that are affected by that modification are modified sequentially.

In combination with other modules, sales order margins can be correctly calculated.

The only "drawback" of using this module is that if a report is obtained by selecting the same date at two different times, the information obtained may be different.

AaronHForgeFlow commented:
Hi @carlosdauden , Thank you for your answer, I will try to reproduce that example. Sometimes I get weird average standard cost in the products for some reason and it is hard to tell why that happens. 

In this case, the correction consist of fixing the price in the stock moves, those prices are propagated to the layers, and finally this module updates the standard price correctly, is that correct?

carlosdauden committed:
[FIX] product_cost_price_avco_sync: Incorrect unit_cost when we have negative stock and vacuum with sale return move
TT33334

pedrobaeza commented:
@carlosdauden please rebase and check Travis for merging this

carlosdauden commented:
Ping @LoisRForgeFlow

sergio-teruel committed:
[FIX] product_cost_price_avco_sync: Create SVL for new stock move lines

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Use regex to extract price from description

chienandalu commented on product_cost_price_avco_sync/models/stock_valuation_layer.py:
```diff
@@ -214,10 +215,12 @@ def process_avco_svl_manual_adjustements(self, svls_dic):
-                svl_dic["value"] = (
-                    standard_price * accumulated_qty
-                ) - accumulated_value
+                match_price = re.findall(r"[+-]?[0-9]+\.[0-9]+\)$", svl.description)
```
Can you add a rationale about this change?

carlosdauden committed:
[IMP] product_cost_price_avco_sync: Compatibility with landed cost

pedrobaeza committed:
[FIX] product_cost_price_avco_sync: Don't fail on missing description

rafaelbn reviewed (approved):
Thank you for this migration! I cannot tested but I guess this has been in production for a long time! 👍

pedrobaeza committed:
[REF] product_cost_price_avco_sync: refactor for having hooks

CarlosRoca13 committed:
[IMP] product_cost_price_avco_sync: Add preprocess_svl_dict

With this dict we are allowed to process some svls before
processing the main avoiding do writes on the method.

TT44663

CarlosRoca13 committed:
[IMP] product_cost_price_avco_sync: Add a way to do some action after processing svls

pedrobaeza reviewed (approved):
Let's merge this one as this has been battle-tested. Further improvements or hooks can be done on successive PRs.

/ocabot merge nobump
